### PR TITLE
feat: CW20 Balance Tracking

### PR DESCRIPTION
**Pull request by Jonathansumner**
Addressing #17

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -132,11 +132,12 @@ type CW20Transfer @entity {
   transaction: Transaction!
   block: Block!
 }
-  
+
 type Account @entity {
   # id is the address
   id: ID!
   nativeBalanceChanges: [NativeBalanceChange]! @derivedFrom(field: "account")
```
Comment (bryanchriswhite) [inline schema.graphql:139]: Can we make this `cw20BalanceChanges`? I think that's a better fit with convention.

Comment (bryanchriswhite) [reply schema.graphql:139]: Can we also try to make the type `Cw20BalanceChange` like we talked about?

File: src/mappings/mappingHandlers.ts [hunk 2]
```diff
@@ -185,7 +185,7 @@ export async function handleCW20Transfer(event: CosmosEvent): Promise<void> {
   const amount = msg.msg?.transfer?.amount;
 
   if (typeof(amount)==="undefined" || typeof(toAddress)==="undefined" || typeof(fromAddress)==="undefined" || typeof(contract)==="undefined") {
-    logger.warn(`[handleCW20Transfer] (${event.tx.hash}): (!SKIPPED!) message is malformed (event.msg.msg.decodedMsg): ${JSON.stringify(event.msg.msg.decodedMsg, null, 2)}`)
+    logger.warn(`[handleCW20Transfer] (${event.tx.hash}): (!SKIPPED!) message is malformed (event.msg.msg.decodedMsg): ${JSON.stringify(msg, null, 2)}`)
```
Comment (bryanchriswhite) [inline src/mappings/mappingHandlers.ts:188]: :ok_hand:

File: src/mappings/mappingHandlers.ts [hunk 3]
```diff
@@ -208,6 +208,72 @@ export async function handleCW20Transfer(event: CosmosEvent): Promise<void> {
...
 
+export async function handleCW20BalanceBurn(event: CosmosEvent): Promise<void> {
+  const id = messageId(event.msg);
+  logger.info(`[handleCW20BalanceBurn] (tx ${event.tx.hash}): indexing CW20BalanceBurn ${id}`);
+
+  const msg = event.msg.msg.decodedMsg;
+  const fromAddress = msg.sender, contract = msg.contract;
+  const amount = msg.msg?.burn?.amount;
+
+  if (typeof(amount)==="undefined" || typeof(fromAddress)==="undefined" || typeof(contract)==="undefined") {
```
Comment (bryanchriswhite) [inline src/mappings/mappingHandlers.ts:219]: I don't think we want the strictness of the `typeof(object) === "undefined` here, as opposed to just `!object`. The difference between the two is that while the former is only true if `object` was `undefined`, the latter is true not only if `object` was `undefined` but also if it was `false`, `null`, `0`, or `""` (all falsy in JS). Since these values are coming from contracts that had to make it through the VM, their values are validated to a certain extent. However, I think it could still be possible for one or more of these values to be falsy but not undefined; for example, what if `amount` was `null`.

TLDR; I think we should check for falsyness instead. Perhaps there's an exception for `""`, in any case where that's a valid value. Perhaps we can also combine this condition with the one that follows.

---

P.S.
If we really needed this strict of a check I we could've factored this out to a helper, for example:
```typescript
function isAnyUndefined(...args: any[]): boolean {
  return args.any(e => typeof(e) === "undefined");
}
```

Then usage would look like:
```typescript
if (isAnyUndefined(amount, fromAddress, contract) {
  ...
}
```

File: src/mappings/mappingHandlers.ts [hunk 4]
```diff
@@ -208,6 +208,72 @@ export async function handleCW20Transfer(event: CosmosEvent): Promise<void> {
   await cw20transfer.save();
 }
 
+export async function handleCW20BalanceBurn(event: CosmosEvent): Promise<void> {
+  const id = messageId(event.msg);
```
Comment (bryanchriswhite) [inline src/mappings/mappingHandlers.ts:212]: I'm thinking we should consider appending "-mint" and "-burn" to the id's for those balance change events for consistency (as we're already doing "-credit" / "-debit" for transfers).

File: src/mappings/mappingHandlers.ts [hunk 5]
```diff
@@ -208,6 +208,72 @@ export async function handleCW20Transfer(event: CosmosEvent): Promise<void> {
...
+
+  if (typeof(amount)==="undefined" || typeof(toAddress)==="undefined" || typeof(contract)==="undefined") {
+    logger.warn(`[handleCW20BalanceMint] (${event.tx.hash}): (!SKIPPED!) message is malformed (event.msg.msg.decodedMsg): ${JSON.stringify(msg, null, 2)}`)
+    return
+  }
+
+  if (!toAddress || !amount || !contract) {
+    logger.warn(`[handleCW20BalanceMint] (tx ${event.tx.hash}): cannot index event (event.event): ${JSON.stringify(event.event, null, 2)}`)
+    return
+  }
```
Comment (bryanchriswhite) [inline src/mappings/mappingHandlers.ts:249]: (same as https://github.com/fetchai/ledger-subquery/pull/106#discussion_r979620627)

File: src/mappings/mappingHandlers.ts [hunk 6]
```diff
@@ -208,6 +208,72 @@ export async function handleCW20Transfer(event: CosmosEvent): Promise<void> {
...
+
+  if (typeof(amount)==="undefined" || typeof(fromAddress)==="undefined" || typeof(toAddress)==="undefined" || typeof(contract)==="undefined") {
+    logger.warn(`[handleCW20BalanceTransfer] (${event.tx.hash}): (!SKIPPED!) message is malformed (event.msg.msg.decodedMsg): ${JSON.stringify(msg, null, 2)}`)
+    return
+  }
+
+  if (!fromAddress || !amount || !contract) {
+    logger.warn(`[handleCW20BalanceTransfer] (tx ${event.tx.hash}): cannot index event (event.event): ${JSON.stringify(event.event, null, 2)}`)
+    return
+  }
```
Comment (bryanchriswhite) [inline src/mappings/mappingHandlers.ts:271]: (same as https://github.com/fetchai/ledger-subquery/pull/106#discussion_r979620627)

File: src/mappings/mappingHandlers.ts [hunk 7]
```diff
@@ -382,22 +448,41 @@ export async function handleNativeBalanceIncrement(event: CosmosEvent): Promise<
...
     blockId: event.block.block.id,
     transactionId: event.tx.hash,
   });
-
   await nativeBalanceChangeEntity.save()
-}
\ No newline at end of file
+}
+
+async function saveCW20BalanceEvent(id: string, address: string, amount: BigInt, contract: string, event: CosmosEvent) {
```
Comment (bryanchriswhite) [inline src/mappings/mappingHandlers.ts:473]: nit: This function signature is quite large. After 4 or so, I tend to prefer consolidating arguments into a data structure. I think an easy way to get most of these under one object would be to use `CW20BalanceChangeEventProps`, which should be importable from the same file as `CW20BalanceChangeEvent`. This type can hold all the values that we're passing to `CW20BalanceChangeEvent.create`. You could also wrap that type in a `Partial` generic to make all of the properties optional (i.e. `Partial<CW20BalanceChangeEventProps>`).

File: test/helpers/field_enums.py [hunk 8]
```diff
@@ -90,6 +90,24 @@ def select_query(cls, table="c_w20_transfers"):
...
+    transaction_id = 5
+    block_id = 6
+
+    @classmethod
+    def select_query(cls, table="c_w20_balance_changes"):
+        return super().select_query(table)
+
+    @classmethod
+    def by_execute_contract_method(cls, method):
+        return f"SELECT c_w20_balance_changes.id, balance_offset, c_w20_balance_changes.contract, account_id FROM c_w20_balance_changes, execute_contract_messages WHERE execute_contract_messages.id = execute_contract_message_id and method = '{method}'"
```
Comment (bryanchriswhite) [inline test/helpers/field_enums.py:108]: It looks like you had to write this out (as opposed to using `cls.select_query`) because you needed to disambiguate the `id` column. I think we could add a boolean argument to `NamedFields.select_query` called `prefix` or something which will append the table name and the dot to the column names. Then we should be able to reduce this further.

**Review by bryanchriswhite — CHANGES_REQUESTED** [review]
Great work @Jonathansumner! :smile: 

I left a couple of comments where I think we should consider changes and a couple more that might help reduce some things.

File: src/mappings/mappingHandlers.ts (hunk 3 above)
Comment (Jonathansumner) [reply src/mappings/mappingHandlers.ts]: I realised that your suggestion to check for falsyness was carried out a few lines beneath - so I removed this unnecessary check

File: test/helpers/field_enums.py [hunk 9]
```diff
@@ -7,8 +7,14 @@ def select_column_names(cls):
...
     @classmethod
-    def select_query(cls, table):
-        return f"SELECT {', '.join(cls.select_column_names())} FROM {table}"
+    def select_query(cls, table, prefix=False):
+        columns = cls.select_column_names()
+        """ More complex queries might require disambiguation, eg. where two 'id' attributes are being referenced
+            - 'relevant_table.id' this prefix would solve it"""
+        if prefix:
+            for column in range(len(columns)):
+                columns[column] = f"{table}.{columns[column]}"
```
Comment (bryanchriswhite) [inline test/helpers/field_enums.py:16]: nit: I think this can be simplified a bit

```suggestion
            columns = [f"{table}.{column}" for column in columns]
```

File: test/helpers/field_enums.py [hunk 10]
```diff
@@ -18,8 +24,8 @@ class BlockFields(NamedFields):
     timestamp = 3
 
     @classmethod
-    def select_query(cls, table="blocks"):
-        return super().select_query(table)
+    def select_query(cls, table="blocks", prefix=False):
```
Comment (bryanchriswhite) [inline test/helpers/field_enums.py:27]: I feel responsible for you having to do this, my apologies. Let's leave it like this for this one but I think in the future it would make more sense if we break `NamedFields` up into 3 module-level functions and call them directly in the enum classes (i.e. stop inheriting).

File: test/helpers/field_enums.py [hunk 11]
```diff
@@ -86,8 +92,8 @@ class CW20TransferFields(NamedFields):
     contract = 7
 
     @classmethod
-    def select_query(cls, table="c_w20_transfers"):
-        return super().select_query(table)
+    def select_query(cls, table="cw20_transfers", prefix=False):
```
Comment (bryanchriswhite) [inline test/helpers/field_enums.py:95]: :partying_face::raised_hands: - so satisfying!

**Review by bryanchriswhite — APPROVED** [review]
Great work @Jonathansumner! :raised_hands: 

I had one more suggestion to simplify your `select_query` changes a tiny bit but it's up to you.

File: test/helpers/field_enums.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jonathansumner) [reply test/helpers/field_enums.py:26]: Hey no worries - it was a quick job with a good result

**Review by bryanchriswhite — APPROVED** [review]
Woot - thanks @Jonathansumner! :partying_face:
